### PR TITLE
User registration

### PR DESCRIPTION
**Pull request by pedbr**
-Adds tests for User registration feature
-Research Devise gem 
-Install Devise gem
-Configure Devise
-Link user registration routes
-Adds Name field as registration requirement

PT> https://www.pivotaltracker.com/story/show/167007919

### CODE REVIEW

File: app/controllers/registrations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class RegistrationsController < Devise::RegistrationsController
+  
+  private
+
+  def sign_up_params
+    params.require(:user).permit(:name, :email, :password, :password_confirmation)
+  end
+
+  def account_update_params
```
Comment (oliverochman) [inline app/controllers/registrations_controller.rb]: is this in scope for this feature?

File: app/views/devise/confirmations/new.html.haml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+%h2 Resend confirmation instructions
```
Comment (oliverochman) [inline app/views/devise/confirmations/new.html.haml]: please remove all devise views that you do not change in any way

File: app/views/products/index.html.haml [hunk 3]
```diff
@@ -3,7 +3,7 @@
     %li
       %a{href: "#"} Cart
     %li
-      %a{href: "#"} Login
+      %a{href: "users/sign_in"} Login
```
Comment (oliverochman) [inline app/views/products/index.html.haml]: Is this for logging in or signing up?

Comment (oliverochman) [reply app/views/products/index.html.haml]: will you be able to see this link when you are logged in?

File: config/application.rb [hunk 4]
```diff
@@ -25,5 +25,15 @@ class Application < Rails::Application
       generate.controller_specs false
       generate.system_tests = false
     end
+    ActionMailer::Base.smtp_settings = {
```
Comment (oliverochman) [inline config/application.rb]: is this in scope for this feature?

File: db/migrate/20190710075926_add_name_to_users.rb [hunk 5]
```diff
@@ -0,0 +1,7 @@
+class AddNameToUsers < ActiveRecord::Migration[4.2]
+  def change
+    add_column :users, :name, :string
+    add_column :users, :date_of_birth, :datetime
```
Comment (oliverochman) [inline db/migrate/20190710075926_add_name_to_users.rb]: why did you add these other columns?

File: features/user_registration.feature [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+  As a user,
+  In order to have be able to finalise an order
+  I would like to be able to register an account
+
+  Background:
+    Given I visit the site
+    And I click the "Login" button
+
+  Scenario: Visitor can create an account to become a registered user[Hapy path]
+    And I click the "Sign up" button
```
Comment (oliverochman) [inline features/user_registration.feature]: Please refactor to 
`When I click the "Sign up" button`

File: features/user_registration.feature [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    Then I fill the "Name" with "John"
```
Comment (oliverochman) [inline features/user_registration.feature]: Please refactor to
`And I fill the "Name" with "John"
`

File: features/user_registration.feature [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    And I fill the "Email" with "john@doe.com"
+    And I fill the "Password" with "password1"
+    And I fill the "Password confirmation" with "password1"
+    And I click the "Sign up" button
+    Then I should see "Hello, John"
+
+  Scenario: Visitor can't create an account with invalid credentials[Sad path]
+    Given I click the "Sign up" button
```
Comment (oliverochman) [inline features/user_registration.feature]: Please refactor to
`When I click the "Sign up" button`

File: features/user_registration.feature [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    And I fill the "Email" with ""
+    And I fill the "Password" with ""
+    And I fill the "Password confirmation" with ""
+    And I click the "Sign up" button
+    Then I should see "Email can't be blank"
+    Then I should see "Password can't be blank"
```
Comment (oliverochman) [inline features/user_registration.feature]: Please refactor to
`And I should see "Password can't be blank" `

File: spec/models/user_spec.rb [hunk 10]
```diff
@@ -0,0 +1,9 @@
+require 'rails_helper'
+
+RSpec.describe User, type: :model do
+  describe 'DB table' do
+    it { is_expected.to have_db_column :name }
+    it { is_expected.to have_db_column :email }
+    it { is_expected.to have_db_column :encrypted_password }
+  end
```
Comment (oliverochman) [inline spec/models/user_spec.rb:8]: You need to test that the FactoryBot is valid here

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: app/controllers/registrations_controller.rb (hunk 1 above)
Comment (pedbr) [reply app/controllers/registrations_controller.rb]: Removed.

File: app/views/devise/confirmations/new.html.haml (hunk 2 above)
Comment (pedbr) [reply app/views/devise/confirmations/new.html.haml]: All views out of this feature's scope removed.

File: app/views/products/index.html.haml (hunk 3 above)
Comment (pedbr) [reply app/views/products/index.html.haml]: Our idea is to have only one button, the "Login", this will direct to the Login view, and in this view there's a sign up button, so if the user doesn't have an account yet he can sign up from there. 

And when a user is logged in, he sees a greeting and a sign out button, we added that in the /views/layouts/application.html.haml >
- if user_signed_in?
    %p
      Hello, #{current_user.name} - #{link_to "Sign_out", destroy_user_session_path, method: 'delete'}

File: config/application.rb (hunk 4 above)
Comment (pedbr) [reply config/application.rb]: Out of scope. Removed.

File: db/migrate/20190710075926_add_name_to_users.rb (hunk 5 above)
Comment (pedbr) [reply db/migrate/20190710075926_add_name_to_users.rb]: We added this when we were trying to add more requirement fields to the sign up form, because Devise only has "email" and "password" and we also wanted to add "name". we watched some tutorials and ended up adding more fields like date of birth, gender and all that. But at the end we are only using the name, so all the excess is now removed.

File: features/user_registration.feature (hunk 6 above)
Comment (pedbr) [reply features/user_registration.feature]: Corrected.

File: features/user_registration.feature (hunk 7 above)
Comment (pedbr) [reply features/user_registration.feature]: Corrected

File: features/user_registration.feature (hunk 8 above)
Comment (pedbr) [reply features/user_registration.feature]: Corrected

File: features/user_registration.feature (hunk 9 above)
Comment (pedbr) [reply features/user_registration.feature]: Corrected

File: spec/models/user_spec.rb (hunk 10 above)
Comment (pedbr) [reply spec/models/user_spec.rb:8]: Factory bot configured and unit tests added and going green

File: app/views/products/index.html.haml (hunk 3 above)
Comment (oliverochman) [reply app/views/products/index.html.haml]: okay fair enough

File: app/views/devise/shared/_links.html.haml [hunk 11]
```diff
@@ -0,0 +1,19 @@
+- if controller_name != 'sessions'
```
Comment (oliverochman) [inline app/views/devise/shared/_links.html.haml]: if you haven't modified this please remove it

File: app/views/devise/shared/_error_messages.html.haml [hunk 12]
```diff
@@ -0,0 +1,9 @@
+- if resource.errors.any?
```
Comment (oliverochman) [inline app/views/devise/shared/_error_messages.html.haml]: if you haven't modified this please remove it

File: app/views/devise/sessions/new.html.haml [hunk 13]
```diff
@@ -0,0 +1,17 @@
+%h2 Log in
```
Comment (oliverochman) [inline app/views/devise/sessions/new.html.haml]: if you haven't modified this please remove it

Comment (oliverochman) [reply app/views/devise/sessions/new.html.haml]: also out of scope

File: app/views/devise/shared/_links.html.haml (hunk 11 above)
Comment (oliverochman) [reply app/views/devise/shared/_links.html.haml]: also out of scope

**Review by oliverochman — CHANGES_REQUESTED** [review]

File: app/views/devise/shared/_links.html.haml (hunk 11 above)
Comment (pedbr) [reply app/views/devise/shared/_links.html.haml]: removed!

File: app/views/devise/shared/_error_messages.html.haml (hunk 12 above)
Comment (pedbr) [reply app/views/devise/shared/_error_messages.html.haml]: We didn't modify it, but this code is what renders the errors when the requirements for the signup is not fulfilled, so we need this for the sad-path

Comment (oliverochman) [reply app/views/devise/shared/_error_messages.html.haml]: But that's the thing with devise and gems like this overall. You will have access to this functionality without having this file in the code base. You only need those views if you need to change anything (overwrite them).  

Try it out without this file

File: app/views/devise/registrations/new.html.haml [hunk 14]
```diff
@@ -0,0 +1,25 @@
+%h2 Sign up
+= form_for(resource, as: resource_name, url: registration_path(resource_name)) do |f|
+  = render "devise/shared/error_messages", resource: resource
+  .field
+    = f.label :name
```
Comment (oliverochman) [inline app/views/devise/registrations/new.html.haml:5]: I'm not complaining about this file here because assume that you needed to add an input field for `:name`

File: app/views/devise/sessions/new.html.haml (hunk 13 above)
Comment (pedbr) [reply app/views/devise/sessions/new.html.haml]: Removed.

File: app/views/devise/registrations/new.html.haml (hunk 14 above)
Comment (pedbr) [reply app/views/devise/registrations/new.html.haml:5]: Yes, we added this to have the extra field for the name.

**Review by oliverochman — APPROVED** [review]
